mcp23017_driver/mcp23017.c

```diff
@@ -0,0 +1,593 @@
+
+    return bit_value;
+}
+
+void set_register_bit(uint8_t* register_value, uint8_t index, uint8_t bit_value) {
+    if (bit_value == 0) {
+        *register_value = *register_value & ~(1 << index);
+    } else if (bit_value == 1) {
+        *register_value = *register_value | (1 << index);
+    }
+}
+
+HAL_StatusTypeDef read_register(MCP23017_t* hdev, uint8_t register_address, uint8_t* register_value) {
+    HAL_StatusTypeDef HAL_Status;
+
+    HAL_Status = HAL_I2C_Mem_Read(
+        hdev->hi2c, 
+        hdev->device_address, 
+        register_address,
+        I2C_MEMADD_SIZE_8BIT, 
+        register_value, 
+        1, 
+        hdev->i2c_timeout
+    );
+
+    return HAL_Status;
+}
+
+HAL_StatusTypeDef write_register(MCP23017_t* hdev, uint8_t register_address, uint8_t register_value) {  
+    HAL_StatusTypeDef HAL_Status;
+    
+    HAL_Status = HAL_I2C_Mem_Write(
+        hdev->hi2c,
+        hdev->device_address,
+        register_address,
+        I2C_MEMADD_SIZE_8BIT,
+        &register_value,
+        1,
+        hdev->i2c_timeout
+    );
+
+    return HAL_Status;
+}
+
+HAL_StatusTypeDef read_register_bit(MCP23017_t* hdev, uint8_t register_address, uint8_t index, uint8_t* register_bit_value) {
+    HAL_StatusTypeDef HAL_Status;
+    uint8_t register_value;
+    
+    HAL_Status = read_register(hdev, register_address, &register_value);
+    *register_bit_value = get_register_bit(register_value, index);
+
+    return HAL_Status;
+}
+
+HAL_StatusTypeDef write_register_bit(MCP23017_t* hdev, uint8_t register_address, uint8_t index, uint8_t register_bit_value) {
+    HAL_StatusTypeDef HAL_Status;
+    uint8_t register_value;
+
+    HAL_Status = read_register(hdev, register_address, &register_value);
+    if (HAL_Status != HAL_OK) {
+        return HAL_Status;
+    }
+
+    set_register_bit(&register_value, index, register_bit_value);
+    HAL_Status = write_register(hdev, register_address, register_value);
+
+    return HAL_Status;
+}
+
+HAL_StatusTypeDef get_config(MCP23017_t* hdev, MCP23017_configuration_t* config) {
+    HAL_StatusTypeDef HAL_Status;
+    uint8_t register_value;
+
+    HAL_Status = read_register(hdev, current_registers_addresses.IOCON, &register_value);
+    if (HAL_Status != HAL_OK) {
+        return HAL_Status;
+    }
+
+    config->bank = get_register_bit(register_value, 7);
+    config->mirror = get_register_bit(register_value, 6);
+    config->seqop = get_register_bit(register_value, 5);
+    config->disslw = get_register_bit(register_value, 4);
+    //config->haen_MCP23S17_only = get_register_bit(register_value, 3);
+    config->odr = get_register_bit(register_value, 2);
+    config->intpol = get_register_bit(register_value, 1);
+    //get_register_bit(register_value, 0);
+
+    return HAL_Status;
+}
+
+HAL_StatusTypeDef set_config(MCP23017_t* hdev, MCP23017_configuration_t config) {
+    HAL_StatusTypeDef HAL_Status;
+    uint8_t register_value;
+
+    set_register_bit(&register_value, 7, config.bank);
```

Comment:
If you could squash all the register bit in one register byte, it would be much better (then setting from 0 to 7 every time)

Reply:
Removed configurations struct. Now the entire register is set every time.